### PR TITLE
Scratch 3

### PR DESCRIPTION
**Pull request by i-make-robots**
so we have a place to comment and make upgrades.

### CODE REVIEW

File: src/main/java/com/marginallyclever/makelangelo/makeArt/io/vector/LoadScratch3.java [hunk 1]
```diff
@@ -446,6 +448,18 @@ private void doMotionSetY(JSONObject currentBlock) throws Exception {
...
+	private void doSetPenColor(JSONObject currentBlock) throws Exception {
+		String c = (String)findInputInBlock(currentBlock,"COLOR");
+		if(c.startsWith("#")) {
+			c = c.substring(1);
+			int r = Integer.parseInt(c.substring(0,2));
+			int g = Integer.parseInt(c.substring(2,4));
+			int b = Integer.parseInt(c.substring(4));
+			logger.debug("SET COLOR {}",c);
+			myTurtle.setColor(new ColorRGB(r,g,b));
+		}
```
Comment (PPAC37) [inline src/main/java/com/marginallyclever/makelangelo/makeArt/io/vector/LoadScratch3.java:460]:
~~~
		double dColor = resolveValue(findInputInBlock(currentBlock,"COLOR"));
		final ColorRGB colorRGB = new ColorRGB(((int)dColor)&0xFFFFFF);
		logger.debug("pen_setPenColorToColor COLOR {}", colorRGB.toString());
		myTurtle.setColor(colorRGB); // TODO find wy i need a move after a setColor befor a penUp/Down or the Makelangelo render dont show anything (but the number of segment is visible in the view slider ... like as the color is store in turtel.x the render take it as a move or someting like that ...)

~~~

and require in `private double resolveValue(Object currentObject) `
the case 
~~~

			case 9:  // color value [9, "#60bf54"] 
				double res = 0;
				try {
					logger.debug("a color {}",currentArray.getString(1));
					ColorRGB color = new ColorRGB(Integer.parseInt(currentArray.getString(1).substring(1), 16) & 0xFFFFFF);
					res = color.toInt();
					logger.debug("a color {}",color.toString());
				}catch (Exception e ){}
				return res;  
			
~~~

But i have some strange result if the color change is not folowed by a move but by a penUp or Down ...

File: src/main/java/com/marginallyclever/makelangelo/makeArt/io/vector/LoadScratch3.java [hunk 2]
```diff
@@ -104,207 +161,384 @@ public Turtle load(InputStream in) throws Exception {
...
+		double px = resolveValue(findInputInBlock(currentBlock,"X"));
+		double py = resolveValue(findInputInBlock(currentBlock,"Y"));
+		logger.debug("GOTO {} {}",px,py);
+		myTurtle.moveTo(px, py);
+	}
+
+	private void doMotionPointInDirection(JSONObject currentBlock) throws Exception {
+		double v = resolveValue(findInputInBlock(currentBlock,"DIRECTION"));
+		logger.debug("POINT AT {}",v);
+		myTurtle.setAngle(v);
```
Comment (PPAC37) [inline src/main/java/com/marginallyclever/makelangelo/makeArt/io/vector/LoadScratch3.java:405]: `myTurtle.setAngle(v-90.0);//a orientation 90° diffrence between Turtle and Scratch`
seem to resolv my orientation issue.

File: src/main/java/com/marginallyclever/makelangelo/makeArt/io/vector/LoadScratch2.java [hunk 3]
```diff
@@ -154,24 +105,81 @@ public Turtle load(InputStream in) throws Exception {
...
+		while((entry = zipInputStream.getNextEntry())!=null) {
+	        if(entry.getName().equals(PROJECT_JSON) ) {
+	        	logger.debug("Found.");
+	        	
+		        // read buffered stream into temp file.
+	        	File tempZipFile = File.createTempFile("project", "json");
+	        	tempZipFile.setReadable(true);
+	        	tempZipFile.setWritable(true);
+	        	tempZipFile.deleteOnExit();
+		        FileOutputStream fos = new FileOutputStream(tempZipFile);
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/makeArt/io/vector/LoadScratch2.java:172]: Use `try with resources` in order to ensure the stream is closed whatever happened, like in https://github.com/MarginallyClever/Makelangelo-software/blob/master/src/main/java/com/marginallyclever/makelangelo/makeArt/io/vector/TurtleFactory.java#L70

File: src/main/java/com/marginallyclever/makelangelo/makeArt/io/vector/LoadScratch2.java [hunk 4]
```diff
@@ -154,24 +105,81 @@ public Turtle load(InputStream in) throws Exception {
...
+	    		byte[] buffer = new byte[2048];
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/makeArt/io/vector/LoadScratch2.java:173]: The code indentation is not homogeneous: there are 4 extra spaces in thoses lines

**Review by coliss86 — CHANGES_REQUESTED** [review]
nice work, some changes should be made

File: src/main/java/com/marginallyclever/makelangelo/makeArt/io/vector/LoadScratch2.java (hunk 3 above)
Comment (i-make-robots) [reply src/main/java/com/marginallyclever/makelangelo/makeArt/io/vector/LoadScratch2.java:172]: doh!  I keep forgetting this, it's new to me.  thank you.

Comment (i-make-robots) [reply src/main/java/com/marginallyclever/makelangelo/makeArt/io/vector/LoadScratch2.java:172]: fixed and also in LoadScratch2

File: src/main/java/com/marginallyclever/makelangelo/makeArt/io/vector/LoadScratch2.java (hunk 4 above)
Comment (i-make-robots) [reply src/main/java/com/marginallyclever/makelangelo/makeArt/io/vector/LoadScratch2.java:173]: fixed and also in LoadScratch2

**Pull request merged**